cartographer_ros/cartographer_ros/metrics/internal/gauge.h

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2018 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CARTOGRAPHER_ROS_METRICS_INTERNAL_GAUGE_H
+#define CARTOGRAPHER_ROS_METRICS_INTERNAL_GAUGE_H
+
+#include <atomic>
+#include <map>
+#include <string>
+
+#include "cartographer/metrics/gauge.h"
+#include "cartographer_ros_msgs/Metric.h"
+
+namespace cartographer_ros {
+namespace metrics {
+
+class Gauge : public ::cartographer::metrics::Gauge {
+ public:
+  explicit Gauge(const std::map<std::string, std::string>& labels)
+      : labels_(labels) {}
+
+  void Decrement(const double by_value) override { Add(-1. * by_value); }
+
+  void Decrement() override { Decrement(1.); }
+
+  void Increment(const double by_value) override { Add(by_value); }
+
+  void Increment() override { Increment(1.); }
+
+  void Set(double value) override {
+    double expected = value_.load();
+    while (!value_.compare_exchange_weak(expected, value)) {
```

Comment:
@MichaelGrupp I'm not sure this is correct. What if two threads get the same "expected", and in one of them compare_exchange_weak succeeds, it returns true, and the while loop ends.
In the other, since the first thread already modified the value, compare_exchange_weak gets into an
infinite loop, until, by chance, someone sets value to the same value it was fetched in load.

Even if I'm misunderstanding, I would highly recommend not trying to be clever with an atomic in this implementation and just using a cartographer::common::Mutex to protect a plain old double value. That's much easier to understand and any potential performance problems would be negligible.

Reply:
@CodeArno I  think this code is correct: when compare_exchange_weak fails, the expected value is updated by a load.

https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange

Reply:
agree with @cschuet